### DISCUSSION TITLE
What are dgroups?

### DISCUSSION THREAD

**Discussion by jgarte** [qtile/qtile#4137] [Q&A]
Hi, what are dynamic groups?

The docstring says that dgroups are [Dynamic groups](https://github.com/qtile/qtile/blob/master/libqtile/dgroups.py#L69) but that doesn't explain what dynamic groups are... ;()

A regular `Group` also says that it is a Dynamic group which is confusing:

https://docs.qtile.org/en/stable/_modules/libqtile/config.html#Group

**Comment by jrwrigh** [comment]
I'm not 100% sure, but that actually steps on what is probably my biggest annoyance with Qtile; client-server proxy structure. 

tl;dr, (I think) the `Group` class is the object that's intended to be interacted with by the user (client-side), while the `DGroups` class is what the server uses to implement the behavior of Groups of windows.

This old comment I wrote up awhile ago explains it well:

Full disclosure: A) It's been quite awhile since I've looked into it. Most of my previous comment comes from trying to do some custom behavior that involves working outside of just the normal configuration commands and options. And since then I figured out a weird workaround to find a "hacking" process that works well for me. B) I don't know a better way of doing it necessarily. I'm not about to try and tell people how to do it better, only where my pain points as a user are. It's possibly those pains are unavoidable C) I absolutely love using qtile. Once I figured a way around the weird roadblocks I was encountering, it's by far my favorite WM and I don't anticipate moving away from it anytime soon. You can love something and still have complaints about it.

So the architecture of qtile is a server-client. The server is a continuously running python process, while the client (via `qtile shell`, or `CommandClient` or `InteractiveCommandClient`) is a separate running python process. To communicate with the server, the client exposes proxy objects of the server that you can interact with and manipulate. This is explained a bit in [the documentation](http://docs.qtile.org/en/stable/manual/commands/index.html) and in great detail [in this GH discussion](https://github.com/qtile/qtile/discussions/2337#discussioncomment-554202)

Now if you're writing functions to do custom behavior, you need to interact with server python-objects. The issue is that the client proxy object and the server object aren't the same and don't behave the same. Most of my grievances are documented in this [Google Groups post](https://groups.google.com/g/qtile-dev/c/-FdY6tQi5P4/m/5I1yzsuuAgAJ). The biggest one imo is that the proxy objects often rely on setting a `__getattr__` method to get the object's methods and properties, rather than setting them explicitly. This means I can't run `dir()` on them and inspect the objects manually.

Note however that I did find a good solution to it via `aiomanhole`, which allows me to directly interact with the server process and inspect/traverse the objects there. See [this config](https://github.com/ramnes/qtile-config/blob/ff02c13e7ddd31a4b964e0d6a301cd0520a85e31/config.py)  from one of the qtile maintainers for how to set up `aiomanhole`.

**Reply by jrwrigh** [reply]
## `Group` vs `DGroups` vs `_Group`

A theoretical "Group" is a collection of windows, where every window belongs to a group and only one group. 

That theoretical collection is most tangibly implemented in the `_Group` class.
`_Group` defines the `CommmandObject` for a single Group. This is effectively the "server-side" object defining a theoretical group.

`DGroups` is a convenience class that helps `Qtile(CommandObject)` manage `_Groups`. 
It spawns `_Group` objects in the `Qtile(CommandObject)` class and creates them based on the configuration (discussed next). It also determines which `_Group` new windows (often called `client`s) are added to based on `Rule`s setup.

`Group` is a configuration class for defining the groups that will be created later. It is primarily used by `DGroups`.

So interacting together, a use wants to define a theoretical "Group". They create a list of `Group` objects in their configuration. That list of `Groups` is then read by `Qtile` from the configuration and handed off to `DGroups` to set them up. `DGroups` goes through the list of `Group`s and spawns them into corresponding `_Group` objects.
